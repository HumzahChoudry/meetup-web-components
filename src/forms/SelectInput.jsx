import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

import Icon from '../media/Icon';

/**
 * @module SelectInput
 */
class SelectInput extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			value: props.value || (props.options[0] || {}).value
		};
		this.onChange = this.onChange.bind(this);
	}

	onChange(e) {
		this.setState({ value: e.target.value });
		if (this.props.onChange) {
			this.props.onChange(e);
		}
	}

	componentWillReceiveProps(nextProps) {
		if (nextProps.value !== this.props.value) {
			this.setState(() => ({ value: nextProps.value }));
		}
	}

	render() {
		const {
			children,
			className,
			labelClassName,
			label,
			options,
			name,
			error,
			errors,
			required,
			onChange, // eslint-disable-line no-unused-vars
			value, // eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'select--reset span--100 padding--selectArrow',
			{ 'field--error': errors && errors.length > 0 || error },
			className
		);

		const labelClassNames = cx(
			'label--field',
			{ required },
			labelClassName
		);

		return (
			<div>
				<div className="inputContainer">
					{label &&
						<label className={labelClassNames} htmlFor={other.id}>
							{label}
						</label>
					}
					<select
						name={name}
						required={required}
						className={classNames}
						onChange={this.onChange}
						value={this.state.value}
						{...other}
					>
						{
							options.map((option, key) =>
								(<option key={key}
									value={option.value}
									disabled={option.disabled}>
									{option.label}
								</option>)
							)
						}
					</select>
					<Icon
						className="select-customArrow"
						shape="chevron-down"
						size="xs"
					/>
				</div>
				{error && <p className='text--error text--small'>{error}</p>}
				{
					errors && errors.length > 0 &&
						errors.map((error, key) =>
							<p key={key} className='text--error text--small'>{error}</p>
						)
				}
				{children}
			</div>
		);
	}
}

SelectInput.propTypes = {
	name: PropTypes.string.isRequired,
	required: PropTypes.bool,
	options: PropTypes.arrayOf(PropTypes.shape({
		label: PropTypes.string.isRequired,
		value: PropTypes.string.isRequired,
	})).isRequired,
	error: PropTypes.string,
	errors: PropTypes.array,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	labelClassName: PropTypes.string,
	value: (props, propName, componentName) => {
		const validValues = props.options.map(opt => opt.value);

		if (props[propName] && !validValues.includes(props[propName])) {
			return new Error(`${propName} prop supplied to ${componentName} does not match any supplied options values`);
		}
	},
};

export default SelectInput;
