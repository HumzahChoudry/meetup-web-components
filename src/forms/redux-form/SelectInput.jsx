import React from 'react';
import SelectInput from '../SelectInput';

/**
 * This component wraps the standard SelectInput component for use with redux-form.
 * It deconstructs props that redux-form sets and re-sets them on SelectInput.
 * @param {object} props React component props
 * @return {React.Component} SelectInput
 */
const ReduxFormSelectInput = props => {
	const { meta, validateBeforeTouched, ...other } = props;

	const error = (validateBeforeTouched || meta.touched) ? meta.error : null;

	return <SelectInput error={error} {...other} />;
};

ReduxFormSelectInput.displayName = 'ReduxFormSelectInput';

export default ReduxFormSelectInput;

