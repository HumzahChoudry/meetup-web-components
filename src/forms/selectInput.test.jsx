import React from 'react';
import TestUtils from 'react-addons-test-utils';
import SelectInput from './SelectInput';

const testOptions = [
	{ label: 'One', value: '1' },
	{ label: 'Two', value: '2' },
	{ label: 'Three', value: '3' }
];
const nameAttribute = 'testSelect';

describe('SelectInput basic', () => {
	let component;

	beforeEach(() => {
		component = TestUtils.renderIntoDocument(
			<SelectInput
				label='Test select'
				name={nameAttribute}
				options={testOptions}
			/>
		);
	});
	afterEach(() => {
		component = null;
	});

	it('exists', () => {
		expect(() => TestUtils.findRenderedDOMComponentWithTag(component, 'select')).not.toThrow();
	});

	it('should have a NAME attribute', () => {
		const selectEl = TestUtils.findRenderedDOMComponentWithTag(component, 'select');
		expect(selectEl.name).toEqual(nameAttribute);
	});

	it('default value should fall back on first option value', () => {
		expect(component.state.value).toBe(testOptions[0].value);
	});

});

describe('SelectInput advanced', () => {
	it('should set correct value in state on change', () => {
		const newValue = '2';
		const onChange = jest.fn();
		const changeSpy = spyOn(SelectInput.prototype, 'onChange').and.callThrough();

		const component = TestUtils.renderIntoDocument(
			<SelectInput
				label='Test select'
				name={nameAttribute}
				options={testOptions}
				onChange={onChange}
			/>
		);
		const selectEl = TestUtils.findRenderedDOMComponentWithTag(component, 'select');

		expect(changeSpy).not.toHaveBeenCalled();
		TestUtils.Simulate.change(selectEl, { target: { value: newValue } });
		expect(changeSpy).toHaveBeenCalled();

		expect(selectEl.value).toEqual(newValue);
	});

	it('should set correct value specified in props', () => {
		const CUSTOM_VALUE = '2';
		const component = TestUtils.renderIntoDocument(
			<SelectInput
				label='Test select'
				name={nameAttribute}
				options={testOptions}
				value={CUSTOM_VALUE}
			/>
		);
		expect(component.state.value).toBe(CUSTOM_VALUE);
	});

	it('should throw error for invalid default value', () => {
		expect(TestUtils.renderIntoDocument(
			<SelectInput
				label='Test select'
				name={nameAttribute}
				options={testOptions}
				value='this is invalid'
			/>
		)).toThrow();
	});
});
