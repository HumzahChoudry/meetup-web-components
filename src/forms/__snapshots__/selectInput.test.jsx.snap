// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput basic renders into the DOM 1`] = `
<div>
  <div
    className="inputContainer"
  >
    <label
      className="label--field"
      htmlFor="testSelect"
    >
      Test select
    </label>
    <select
      aria-describedby="testSelect-error"
      aria-invalid={false}
      className="select--reset span--100 padding--selectArrow"
      id="testSelect"
      name="testSelect"
      onChange={[Function]}
      value="1"
    >
      <option
        key="0"
        value="1"
      >
        One
      </option>
      <option
        key="1"
        value="2"
      >
        Two
      </option>
      <option
        key="2"
        value="3"
      >
        Three
      </option>
      <option
        disabled={true}
        key="3"
        value="4"
      >
        Four
      </option>
    </select>
    <Icon
      className="select-customArrow"
      shape="chevron-down"
      size="xs"
    />
  </div>
  <ErrorList
    errorId="testSelect-error"
  />
</div>
`;
